### PR TITLE
Chore update golem messages from v2.17.0 to v2.22.0

### PR DESCRIPTION
**Pull request by kbeker**
There is no issue for that

### CODE REVIEW

File: concent_api/core/constants.py [hunk 1]
```diff
@@ -90,3 +90,6 @@ class VerificationResult(IntEnum):
 
 # Defines a string that is returned from Bankster to Core views to indicate timestamp error case.
 BANKSTER_TIMESTAMP_ERROR = 'timestamp_error'
+
+
+ZERO_SIGNATURE = b'\x00' * 65
```
Comment (rwrzesien) [inline concent_api/core/constants.py]: Isn't this available in golem-messages? If yes, maybe it would be better to import it from there.

File: concent_api/core/message_handlers.py [hunk 2]
```diff
@@ -832,7 +832,7 @@ def handle_send_force_payment(
             payment_ts              = parse_datetime_to_timestamp(claim_against_requestor.closure_time),
             task_owner_key          = requestor_ethereum_public_key,
             provider_eth_account    = provider_eth_address,
-            amount_paid             = claim_against_requestor.amount,
+            amount_paid             = int(claim_against_requestor.amount),
```
Comment (rwrzesien) [inline concent_api/core/message_handlers.py]: Please add property to `DepositClaim` like `amount_as_int` which will do the converting and use it.

File: concent_api/concent_api/tests/test_shutdown_mode.py [hunk 3]
```diff
@@ -158,7 +158,7 @@ def test_in_soft_shutdown_mode_concent_should_not_accept_payment_requests(self):
         subtask_results_accepted_list = [
             self._get_deserialized_subtask_results_accepted(
                 timestamp="2018-02-05 10:00:15",
-                payment_ts="2018-02-05 12:00:16",
+                payment_ts="2018-02-05 9:00:16",
```
Comment (rwrzesien) [inline concent_api/concent_api/tests/test_shutdown_mode.py]: Why this change?

File: concent_api/core/message_handlers.py [hunk 4]
```diff
@@ -1142,8 +1142,8 @@ def handle_messages_from_database(client_public_key: bytes) -> Union[message.Mes
                 payment_ts=parse_datetime_to_timestamp(payment_message.payment_ts),
                 task_owner_key=payment_message.task_owner_key.tobytes(),
                 provider_eth_account=payment_message.provider_eth_account,
-                amount_paid=payment_message.amount_paid,
-                amount_pending=payment_message.amount_pending,
+                amount_paid=int(payment_message.amount_paid),
```
Comment (rwrzesien) [inline concent_api/core/message_handlers.py]: Same here, add properties to `PaymentInfo` model.

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py [hunk 5]
```diff
@@ -21,9 +23,16 @@ class NonceAbstractMessage(Message):
         'nonce',
     ] + Message.__slots__
 
+    def slots(self) -> List[List[Union[str, bytes, int]]]:
+        slots = super(NonceAbstractMessage, self).slots()
```
Comment (rwrzesien) [inline middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: You can call `super()` without parameters in Python3.

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py [hunk 6]
```diff
@@ -21,9 +23,16 @@ class NonceAbstractMessage(Message):
...
+        if ['nonce', self.nonce] in slots:  # pylint: disable=no-member
+            slots[slots.index(['nonce', self.nonce])][1] = str(self.nonce)  # pylint: disable=no-member
```
Comment (rwrzesien) [inline middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: Is this really necessary? Maybe it is OK to keep internally `nonce` as int, and add here property too for getting is as string? But if there is a case that it is required for this conversion to happen in `slots()` please let me know.

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py [hunk 7]
```diff
@@ -21,9 +23,16 @@ class NonceAbstractMessage(Message):
...
+        return slots
+
     def deserialize_slot(self, key: str, value: Any) -> Any:
         value = super().deserialize_slot(key, value)
         if key == 'nonce':
+            value = int(value)
```
Comment (rwrzesien) [inline middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: If you apply comment above you can remove this.

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py [hunk 8]
```diff
@@ -49,6 +58,16 @@ class TransactionAbstractMessage(NonceAbstractMessage):
         'data',
     ] + NonceAbstractMessage.__slots__
 
+    def slots(self) -> List[List[Union[str, bytes, int]]]:
+        slots = super(TransactionAbstractMessage, self).slots()
```
Comment (rwrzesien) [inline middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: Same here, why is it required to convert those values?

File: concent_api/core/models.py [hunk 9]
```diff
@@ -497,13 +497,15 @@ def clean(self) -> None:
...
 
         # If available, the report_computed_task nested in force_get_task_result must match report_computed_task.
-        if (
-            self.force_get_task_result is not None and
-            deserialize_message(self.force_get_task_result.data).report_computed_task != deserialized_report_computed_task
-        ):
-            raise ValidationError({
-                'force_get_task_result': "ReportComputedTask nested in ForceGetTaskResult must match Subtask's ReportComputedTask."
-            })
+        if self.force_get_task_result is not None:
```
Comment (pawelkisielewicz) [inline concent_api/core/models.py]: Please look at the comment above. In your version it does not match to new block of code in my opinion. 
Second thing- I know that `clean` is very long method now and you did it similar to the rest, but please extract your new block of code to one inner method (e.g. `_deserialize_force_get_task_result`) and do the whole logic in this method.

File: concent_api/core/tests/test_auth_accept_or_reject_results.py [hunk 10]
```diff
@@ -443,7 +443,7 @@ def test_requestor_sends_subtask_results_accepted_and_concent_should_return_it_t
             timestamp="2018-02-05 10:00:43",
             subtask_results_accepted=self._get_deserialized_subtask_results_accepted(
                 timestamp="2018-02-05 10:00:43",
-                payment_ts="2018-02-05 10:00:44",
+                payment_ts="2018-02-05 9:00:44",
```
Comment (pawelkisielewicz) [inline concent_api/core/tests/test_auth_accept_or_reject_results.py]: In tests there is lot of changes like this - why is it changed and why exactly one hour?

File: concent_api/core/constants.py (hunk 1 above)
Comment (pawelkisielewicz) [reply concent_api/core/constants.py]: What is this signature?

File: concent_api/core/tests/test_integration_force_payment.py [hunk 11]
```diff
@@ -49,7 +49,7 @@ def test_provider_send_force_payment_with_subtask_results_accepted_signed_by_dif
             ),
             self._get_deserialized_subtask_results_accepted(
                 timestamp="2018-02-05 9:00:15",
-                payment_ts="2018-02-05 11:00:16",
+                payment_ts="2018-02-05 8:00:16",
```
Comment (pawelkisielewicz) [inline concent_api/core/tests/test_integration_force_payment.py]: why changed and why 3 hours? In other place I've seen also 2 hours

File: concent_api/core/tests/test_payments_bankster.py [hunk 12]
```diff
@@ -417,14 +417,14 @@ def test_that_settle_overdue_acceptances_should_return_claim_deposit_with_amount
...
-                    payment_ts="2018-02-04 10:00:05",
+                    payment_ts="2018-02-04 9:00:05",
                 ),
                 self._get_deserialized_subtask_results_accepted(
                     report_computed_task=self._get_deserialized_report_computed_task(
                         timestamp="2018-02-04 23:00:05",
                         task_to_compute=task_to_compute
                     ),
-                    payment_ts="2018-02-05 10:00:05",
+                    payment_ts="2018-02-04 10:00:05",
```
Comment (pawelkisielewicz) [inline concent_api/core/tests/test_payments_bankster.py]: Is this change necessary and why? BTW, timestamp of created above `report_computed_task` is 2018-02-04 23:00:05, so I think it is wrong to set `payment_ts` as You did.. It should be later

File: concent_api/core/models.py (hunk 9 above)
Comment (Jakub89) [reply concent_api/core/models.py]: Also, in previous version, if `self.force_get_task_result` was in fact `None`, we've raised `ValidationError`. Now it's just pass, is that a bug or we now want to allow `None` value?

File: concent_api/core/message_handlers.py (hunk 4 above)
Comment (kbeker) [reply concent_api/core/message_handlers.py]: Done

File: concent_api/core/message_handlers.py (hunk 2 above)
Comment (kbeker) [reply concent_api/core/message_handlers.py]: Done

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py (hunk 5 above)
Comment (kbeker) [reply middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: Done

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py (hunk 6 above)
Comment (kbeker) [reply middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: In one of few last updates golem_messages Golem Team changed `cbor2` library to `cbor`, because it's faster than `cbor2`. This library has issues with serialization big integers and that's why we need to convert it during serialization and opposite thing during deserialization. We in our code don't need `nonce`, `v`, `r` or `s` as a string, but this serialization needs it.

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py (hunk 8 above)
Comment (kbeker) [reply middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: Answered above https://github.com/golemfactory/concent/pull/1050#discussion_r245930285

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py (hunk 7 above)
Comment (kbeker) [reply middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: Added comment in code

File: concent_api/core/tests/test_payments_bankster.py (hunk 12 above)
Comment (kbeker) [reply concent_api/core/tests/test_payments_bankster.py]: Newest `golem_messages` does not allow to create `SubtaskResultsAccepted` with `payment_ts` from future. That's why here and in all others `payment_ts` changes it is changed that way.

File: concent_api/core/tests/test_integration_force_payment.py (hunk 11 above)
Comment (kbeker) [reply concent_api/core/tests/test_integration_force_payment.py]: Answered in https://github.com/golemfactory/concent/pull/1050#discussion_r245933221

File: concent_api/core/tests/test_auth_accept_or_reject_results.py (hunk 10 above)
Comment (kbeker) [reply concent_api/core/tests/test_auth_accept_or_reject_results.py]: Answered in https://github.com/golemfactory/concent/pull/1050#discussion_r245933221

File: concent_api/core/models.py (hunk 9 above)
Comment (kbeker) [reply concent_api/core/models.py]: @pawelkisielewicz 
> Please look at the comment above. In your version it does not match to new block of code in my opinion.

This block of code has same rules of working like before. Only one difference in this is that we can deserialize it if `self.force_get_task_result.data` is database object instead of string of bytes.

>Second thing- I know that clean is very long method now and you did it similar to the rest, but please extract your new block of code to one inner method (e.g. _deserialize_force_get_task_result) and do the whole logic in this method.

I think that this issue and this branch is not good place for refactor like this. It could be added new issue for refactor `clean()` method in `Subtask` model. If we want and we have time for this we can refactor whole `Subtask` not just a part of it.

@Jakub89 
>Also, in previous version, if self.force_get_task_result was in fact None, we've raised ValidationError. Now it's just pass, is that a bug or we now want to allow None value?

Previously we raised `ValidationError` only when `self.force_get_task_result is not None` **and** nested `report_computed_task of correctly deserialized message was different than `report_computed_task` stored separately. This block of code still works the same:
- we check if `self.force_get_task_result` exists
- try to deserialize it
- compare `self.force_get_task_results.report_computed_task` with `report_computed_task` of this same `Subtask` but stored separately and if it's not the same we raise `ValidationError`

File: concent_api/core/constants.py (hunk 1 above)
Comment (pawelkisielewicz) [reply concent_api/core/constants.py]: seems strange for me... but if is is so - :+1:

File: concent_api/core/tests/test_payments_bankster.py (hunk 12 above)
Comment (pawelkisielewicz) [reply concent_api/core/tests/test_payments_bankster.py]: OK.. But still we have freeze_time("**2018-02-04 10:00:15**") and timestamp="**2018-02-04 23:00:05**" in one of `report_computed_task`. This should fail for me, how can someone send at **10:00:15** `report_computed_task` with timestamp "**23:00:05**" (both in the same day). 
@kbeker You didn't change this times, so thanks for explanation. :+1: 
Should we validate such situations or leave it as it is now? @cameel ?

File: concent_api/core/models.py (hunk 9 above)
Comment (pawelkisielewicz) [reply concent_api/core/models.py]: >I think that this issue and this branch is not good place for refactor like this. It could be added new issue for refactor clean() method in Subtask model. If we want and we have time for this we can refactor whole Subtask not just a part of it.
I disagree, extraction some code (which is changed by You) to one method is not refactor. If You are touching some code and there is possibility to leave it better I think You should do it.

File: concent_api/core/tests/test_payments_bankster.py (hunk 12 above)
Comment (cameel) [reply concent_api/core/tests/test_payments_bankster.py]: `RCT.timestamp > SRA.payment_ts` would be weird but it should be very rare because it would require one of them to be in the future or in the past when sent. Clients validate that received messages are not too new or too old so only a modified client could do that. And with a modified client our only concern is whether it can break something when it's malicious.

I don't think it breaks anything. We have safeguards against `payments_ts` too far in the past where it matters (e.g. #1036). RCT cannot be too far in the future either because it must be earlier than the deadline.

Such weird situations with timestamps are very often possible with nested messages. My approach here is not to try to validate all the relationships (that would be best but it's too complex because we can never treat timestamps as exact and there are so many of them). Just validate the things that can cause problems.

File: concent_api/core/constants.py [hunk 13]
```diff
@@ -90,3 +91,6 @@ class VerificationResult(IntEnum):
 
 # Defines a string that is returned from Bankster to Core views to indicate timestamp error case.
 BANKSTER_TIMESTAMP_ERROR = 'timestamp_error'
+
+
+ZERO_SIGNATURE = b'\0' * Message.SIG_LEN
```
Comment (dybi) [inline concent_api/core/constants.py]: it is used only in the tests - it shouldn't be here... Please move it to a new modul `tests_constants` or `constants_for_tests`

File: concent_api/core/models.py [hunk 14]
```diff
@@ -497,13 +497,15 @@ def clean(self) -> None:
...
+            if isinstance(self.force_get_task_result.data, bytes):
```
Comment (dybi) [inline concent_api/core/models.py]: why this new `if ... else`? When it would be possible for `self.force_get_task_result.data` not to be `bytes`?

File: concent_api/core/tests/test_core_views.py [hunk 15]
```diff
@@ -190,17 +190,22 @@ def test_send_should_return_http_400_if_data_is_incorrect(self):
...
+        )
+        malformed_dumped_message = dump(
                 data,
                 self.PROVIDER_PRIVATE_KEY,
-                CONCENT_PUBLIC_KEY),
+                CONCENT_PUBLIC_KEY,
+        )
+        response = self.send_request(
+            url='core:send',
+            data=malformed_dumped_message[:-1],
```
Comment (dybi) [inline concent_api/core/tests/test_core_views.py]: please move this slicing (`[-:1]` to the variable definition - otherwise the name is wrong ;)

File: concent_api/core/tests/test_core_views.py [hunk 16]
```diff
@@ -477,22 +482,30 @@ def test_send_task_to_compute_without_public_key_should_return_http_400(self):
...
+                report_computed_task=message.tasks.ReportComputedTask(
+                    task_to_compute=self.task_to_compute
+                ),
+            ),
+            privkey=self.PROVIDER_PRIVATE_KEY,
+            pubkey=settings.CONCENT_PUBLIC_KEY,
+        )
         with mock.patch('core.decorators.log') as log:
-            response = self._send_force_report_computed_task()
+            response = self._send_force_report_computed_task(serialized_force_report_computed_task=dumped_message[:-1])
```
Comment (dybi) [inline concent_api/core/tests/test_core_views.py]: please move this slicing (`[-:1]` to the variable definition - otherwise the name is wrong ;)

File: concent_api/core/tests/test_integration_force_payment.py [hunk 17]
```diff
@@ -231,65 +231,6 @@ def test_that_if_provider_sends_force_payment_with_two_different_provider_ethere
             }
         )
 
-    def test_provider_send_force_payment_beyond_payment_time_concent_should_reject(self):
```
Comment (dybi) [inline concent_api/core/tests/test_integration_force_payment.py]: this was removed because now GM don't allow to create such messages with wrong `payment_ts` or what was the reason?

File: concent_api/core/tests/test_integration_report_computed_task.py [hunk 18]
```diff
@@ -1835,54 +1834,6 @@ def test_provider_forces_computed_task_report_empty_key_returns_400_error(self):
 
         self._assert_client_count_is_equal(0)
 
-    def test_requestor_sends_ack_report_computed_task_with_message_cannot_compute_task(self):
```
Comment (dybi) [inline concent_api/core/tests/test_integration_report_computed_task.py]: and this? another validation in GM?

File: concent_api/core/tests/test_payments_bankster.py (hunk 12 above)
Comment (dybi) [reply concent_api/core/tests/test_payments_bankster.py]: I agree with @pawelkisielewicz - such a timestamp of the 2nd `report_computed_task` doesn't make sense... please modify it to be more reasonable

File: concent_api/core/tests/test_task_verification_result.py [hunk 19]
```diff
@@ -201,7 +203,7 @@ def setUp(self):
         (self.PROVIDER_PRIVATE_KEY, self.PROVIDER_PUBLIC_KEY) = generate_ecc_key_pair()
         (self.REQUESTOR_PRIVATE_KEY, self.REQUESTOR_PUBLIC_KEY) = generate_ecc_key_pair()
 
-        task_to_compute = TaskToComputeFactory()
+        task_to_compute = TaskToComputeFactory(sig=ZERO_SIGNATURE)
```
Comment (dybi) [inline concent_api/core/tests/test_task_verification_result.py]: why do we need this?

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py [hunk 20]
```diff
@@ -49,6 +61,18 @@ class TransactionAbstractMessage(NonceAbstractMessage):
         'data',
     ] + NonceAbstractMessage.__slots__
 
+    def slots(self) -> List[List[Union[str, bytes, int]]]:
+        # cbor library has issuses with big integers serialization
+        # before serialization every big integer need to be convert into string
+        slots = super().slots()
+        if ['gasprice', self.gasprice] in slots:  # pylint: disable=no-member
```
Comment (dybi) [inline middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: I would be for some kind of conversion map, instead of all this `if`s ;)

File: middleman_protocol/middleman_protocol/concent_golem_messages/message.py [hunk 21]
```diff
@@ -114,9 +140,23 @@ class SignedTransaction(TransactionAbstractMessage):
         's',
     ] + TransactionAbstractMessage.__slots__
 
+    def slots(self) -> List[List[Union[str, bytes, int]]]:
+        # cbor library has issuses with big integers serialization
+        # before serialization every big integer need to be convert into string
+        slots = super().slots()
+        if ['v', self.v] in slots:  # pylint: disable=no-member
```
Comment (dybi) [inline middleman_protocol/middleman_protocol/concent_golem_messages/message.py]: same here - map would be much better in my opinion

File: concent_api/core/constants.py (hunk 13 above)
Comment (kbeker) [reply concent_api/core/constants.py]: Done

File: concent_api/core/models.py (hunk 14 above)
Comment (kbeker) [reply concent_api/core/models.py]: Because there are some cases when `self.force_get_task_result.data` could be `StoredMessage` object instead of `bytes`. We have same situation with `self.report_computed_task.data`

File: concent_api/core/tests/test_core_views.py (hunk 15 above)
Comment (kbeker) [reply concent_api/core/tests/test_core_views.py]: Changed :)

File: concent_api/core/tests/test_core_views.py (hunk 16 above)
Comment (kbeker) [reply concent_api/core/tests/test_core_views.py]: Changed :)

File: concent_api/core/tests/test_integration_force_payment.py (hunk 17 above)
Comment (kbeker) [reply concent_api/core/tests/test_integration_force_payment.py]: That's right, GM now does not allow to create messages with `payment_ts` from future

File: concent_api/core/tests/test_integration_report_computed_task.py (hunk 18 above)
Comment (kbeker) [reply concent_api/core/tests/test_integration_report_computed_task.py]: Yup, GM now contains validation for wrong nested message.

File: concent_api/core/tests/test_task_verification_result.py (hunk 19 above)
Comment (kbeker) [reply concent_api/core/tests/test_task_verification_result.py]: @dybi 
When we store to database single message then during serialization its gets own signature - if we serialialize it without any keys then to message is assign `fake_sign'. Let me show you how it's work. I'm gonna ignore parameters which we don't need here e.g: 
```python
store_subtask(
                task_to_compute=task_to_compute,
                report_computed_task=report_computed_task,
            )
```

We are here storing some `Subtask` with `task_to_compute` and `report_compute_task` which has as a nested message `task_to_compute`. When we serialize `task_to_compute` it will get `sig=fake_sig`. Same situation will be with `report_computed_task` beside `task_to_compute` which is nested in RCT. Nested message won't get its own `fake_sig` and that's why it will be different then `task_to_compute` stored separately. When `Subtask` model runs `clean()` method it will get `ValidationError` because `task_to_compute != report_computed_task.task_to_cumpute`. This is the only reason why we have to add `ZERO_SIGNATURE` here. It won't happen in normal cases because `TaskToCompute` is always signed.
